### ISSUE TITLE
[BUG] Unable to GROMACS force field parameters from Slipids_2020

### ISSUE THREAD

**Issue by ale94mleon** [OpenBioSim/sire#26]
**Describe the bug**
`bss.IO.readMolecules` is not working as excpected on lipid molecules. 

**To Reproduce**
1. Copy the files in the working directory (they are one POPC lipd and its corresponded GROMACS topology files): 
[files.zip](https://github.com/OpenBioSim/biosimspace/files/10992689/files.zip). I am Using [Slipids2020 force field](https://zenodo.org/record/4194571)
 
3. Run the code

```python
import BioSimSpace as bss
import sire

try:
    membrane = bss.IO.readMolecules(["conf.gro", "topol.top"])
except:
    print(sire.error.get_last_error_details())
```
3. This is the exception that was raised:

> {'where': 'FILE: /usr/share/miniconda3/envs/sire_build/conda-bld/sire_1675934655021/work/wrapper/IO/_IO_load.cpp, LINE: 160, FUNCTION: SireSystem::System load_molecules(const QStringList&, const SireBase::PropertyMap&)', 'why': 'Cannot load the molecules: There is not enough information in this parser (Gro87( title() = Protein in water, nAtoms() = 134, nResidues() = 1, nFrames() = 1, hasCoordinates() = 1, hasVelocities() = 0 )) to start the creation of a new System. You need to use a more detailed input file.', 'pid': '', 'type': 'SireError::io_error', 'backtrace': 'Backtrace disabled. Call sire.error.enable_backtrace_exceptions() to re-enable.', 'from': 'FILE: /usr/share/miniconda3/envs/sire_build/conda-bld/sire_1675934655021/work/wrapper/IO/_IO_load.cpp, LINE: 160, FUNCTION: SireSystem::System load_molecules(const QStringList&, const SireBase::PropertyMap&)'} 

**Expected behavior**
Load the molecule

**(please complete the following information):**
 - OS: Linux (distribution):
 `cat /etc/*-release`:
```bash
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://www.archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux
```
 - Version of Python: 3.9.16
 - Version of BioSimSpace: `sire --version`: `Sire version: 2023.1.2 (dc4c1cd42953583375f162f5f240169d568ef6a9/Christopher Woods/2023-02-09T09:22:32+00:00 | **UNCLEAN**)`
 - I confirm that I have checked this bug still exists in the latest released version of BioSimSpace: [yes]

**Comment by chryswoods** [comment]
Thanks Lester :-)

**Comment by chryswoods** [comment]
Yes - that is exactly what I see. We need to add some functionality to sire to make it easier to add a path to additional gromacs forcefield files (I did it by setting the `GROMACS_HOME` environment variable to point to the untarred directory).

Then, when it loads, I get the same errors related to Urey-Bradley parameters (missing functionality in sire that needs adding) plus the missing dihedrals (something weird going on in the file? - needs debugging).

This can be assigned to me to fix. It may take about a week.

**Comment by chryswoods** [comment]
Here's my script

```
In [3]: import os

In [4]: os.environ["GROMACS_HOME"] = "forcefield"

In [5]: sr.load("topol.top", format="GROTOP")
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
│                                                                                                  │
│ /Users/chris/mambaforge/envs/openbiosim/lib/python3.10/site-packages/sire/_load.py:409 in load   │
│                                                                                                  │
│   406 │                                                                                          │
│   407 │   map = {"GROMACS_PATH": _get_gromacs_dir(), "show_warnings": show_warnings}             │
│   408 │                                                                                          │
│ ❱ 409 │   return load_molecules(paths, map=create_map(map))                                      │
│   410                                                                                            │
│   411                                                                                            │
│   412 def _to_legacy_system(molecules):                                                          │
│                                                                                                  │
│ /Users/chris/mambaforge/envs/openbiosim/lib/python3.10/site-packages/sire/io/__init__.py:43 in   │
│ load_molecules                                                                                   │
│                                                                                                  │
│   40 │   from ..legacy.IO import load_molecules as _load_molecules                               │
│   41 │   from ..system import System                                                             │
│   42 │                                                                                           │
│ ❱ 43 │   return System(_load_molecules(*args, **kwargs))                                         │
│   44                                                                                             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
OSError: SireError::io_error: Cannot load the molecules: Could not construct a molecule system from the information stored in this Gromacs topology file. Errors include:
Error constructing the molecule associated with template 'Other' : Error getting angle properties for Other. SireError::incompatible_error: Cannot convert a Urey-Bradley 
Gromacs angle into an expression of only the angle size if the UB bond force constant is non-zero - GromacsAngle( functionType() = Urey-Bradley, parameters() = [ 109.5, 502.08,
0.2466, 21756.8 ] )
There were warnings parsing this molecule template:
Ignoring 356 'pairs' lines
e.g. the first ignored pairs line is
1 18 1
Cannot find the dihedral parameters for the dihedral between atoms GroAtom( name() = C31, number() = 40 )-GroAtom( name() = O31, number() = 39 )-GroAtom( name() = C32, number()
= 42 )-GroAtom( name() = O32, number() = 41 ) (atom types CL-OSL-CTL2-OBL, function type 2).
Cannot find the dihedral parameters for the dihedral between atoms GroAtom( name() = C21, number() = 31 )-GroAtom( name() = O21, number() = 30 )-GroAtom( name() = C22, number()
= 33 )-GroAtom( name() = O22, number() = 32 ) (atom types CL-OSL-CTL2-OBL, function type 2). (call sire.error.get_last_error_details() for more info)
```

(I renamed the untarred directory `forcefield`)

**Comment by lohedges** [comment]
Thanks @chryswoods. Yes, I think it would be good to document `GROMACS_HOME` more clearly, or add an option to `readMolecules` to pass it in.

**Comment by ale94mleon** [comment]
Hi, thanks for your quick replay! Indeed the Slipid2020 force field that I am using it is not exactly the same as the one in the link that I provided in my first post (the modifications are really small). And the files `top` file was generated by another script, probably for that reason the problem with the paths.
Therefore, for further debugs, I think that it may be useful a more straightforward example, and also for people that read this post in the future. Here are all the necessary files: [all_files.zip](https://github.com/OpenBioSim/sire/files/11000801/all_files.zip).

Then, it is just to run:

```python
import tarfile
from zipfile import ZipFile
import subprocess
import BioSimSpace as bss
import sire


# Extracting files
with ZipFile('all_files.zip', 'r') as zip:
    zip.extractall()
tar = tarfile.open('Slipids_2020.ff.tar.gz', 'r:gz')
tar.extractall()
tar.close()

# Run the GROMACS command
command = "gmx pdb2gmx -f mem_1POPC.pdb -ff Slipids_2020 -water none -o conf.gro -p topol.top -i posre.itp"
process = subprocess.run(command, shell = True, executable = '/bin/bash', stdout=subprocess.PIPE, stderr=subprocess.PIPE, text = True)
returncode = process.returncode
if returncode != 0:
    print(f'Command {command} returned non-zero exit status {returncode}')
    raise RuntimeError(process.stderr)

try:
    membrane = bss.IO.readMolecules(["conf.gro", "topol.top"])
except:
    print(sire.error.get_last_error_details())
```

What gives:

> INFO:numexpr.utils:Note: NumExpr detected 24 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
> INFO:numexpr.utils:NumExpr defaulting to 8 threads.
> WARNING:root:Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
> {'pid': '', 'from': 'FILE: /usr/share/miniconda3/envs/sire_build/conda-bld/sire_1675934655021/work/wrapper/IO/_IO_load.cpp, LINE: 160, FUNCTION: SireSystem::System load_molecules(const QStringList&, const SireBase::PropertyMap&)', 'where': 'FILE: /usr/share/miniconda3/envs/sire_build/conda-bld/sire_1675934655021/work/wrapper/IO/_IO_load.cpp, LINE: 160, FUNCTION: SireSystem::System load_molecules(const QStringList&, const SireBase::PropertyMap&)', 'backtrace': 'Backtrace disabled. Call sire.error.enable_backtrace_exceptions() to re-enable.', 'why': "Cannot load the molecules: Could not construct a molecule system from the information stored in this Gromacs topology file. Errors include:\nError constructing the molecule associated with template 'Other' : Error getting angle properties for Other. SireError::incompatible_error: Cannot convert a Urey-Bradley Gromacs angle into an expression of only the angle size if the UB bond force constant is non-zero - GromacsAngle( functionType() = Urey-Bradley, parameters() = [ 110.1, 221.752, 0.2179, 18853.1 ] )\nThere were warnings parsing this molecule template:\nIgnoring 356 'pairs' lines\ne.g. the first ignored pairs line is\n1 18 1\nCannot find the dihedral parameters for the dihedral between atoms GroAtom( name() = C21, number() = 31 )-GroAtom( name() = O21, number() = 30 )-GroAtom( name() = C22, number() = 33 )-GroAtom( name() = O22, number() = 32 ) (atom types CL-OSL-CTL2-OBL, function type 2).\nCannot find the dihedral parameters for the dihedral between atoms GroAtom( name() = C31, number() = 40 )-GroAtom( name() = O31, number() = 39 )-GroAtom( name() = C32, number() = 42 )-GroAtom( name() = O32, number() = 41 ) (atom types CL-OSL-CTL2-OBL, function type 2).", 'type': 'SireError::io_error'}

**Comment by lohedges** [comment]
Many thanks for this. It is great to have a self-contained example.

**Comment by chryswoods** [comment]
I am making progress now in [this branch](https://github.com/openbiosim/sire/tree/fix_26). I have added support for Urey Bradley terms to our Gromacs topology parser and have confirmed this works. However, I am still working on the missing parameters that are in the file.

I can parse if I comment out these lines in `topol.top`

```
[ dihedrals ]
;  ai    aj    ak    al funct            c0            c1            c2            c3
;   31    30    33    32     2
;   40    39    42    41     2
```

If I leave them uncommented then I see the same missing parameters error reported above, e.g.

```
Error constructing the molecule associated with template 'Other' : Cannot find the dihedral parameters for the dihedral between atoms 
GroAtom( name() = C31, number() = 40 )-GroAtom( name() = O31, number() = 39 )-GroAtom( name() = C32, number() = 42 )-GroAtom( name() = 
O32, number() = 41 ) (atom types CL-OSL-CTL2-OBL, function type 2).
Cannot find the dihedral parameters for the dihedral between atoms GroAtom( name() = C21, number() = 31 )-GroAtom( name() = O21, 
number() = 30 )-GroAtom( name() = C22, number() = 33 )-GroAtom( name() = O22, number() = 32 ) (atom types CL-OSL-CTL2-OBL, function type
2)
```

The error is because the parser can't find the dihedrals for atom types `CL-OSL-CTL2-OBL` (both dihedrals need the parameters defined by these types)

**Comment by chryswoods** [comment]
Ok - the parameter for this dihedral exists - it is just that sire isn't finding it. The parameter is at the end of `Slipids_2020.ff/ffbonded.itp`, i.e.

```
[ dihedraltypes ]
; i     j       k       l       func    q0      cq
OBL     X       X       CL      2       0.00    836.8
OCL     X       X       CCL     2       0.00    803.328 
NH1     X       X       H11     2       0.0000  167.36
```

It is the `OBL X X CL` parameter. I'll look now to find out why this isn't being found by sire...

**Comment by chryswoods** [comment]
I found the issue - I had not added code to do wildcard searching for gromacs dihedrals using `A-*-*-D` or `D-*-*-A`. I've added this in the [latest commit](https://github.com/OpenBioSim/sire/commit/654635dcf397bbcc7cf43cbdf34fcb7b44aab0b9).

This then raised another issue which is that this was actually an improper term, with the functional form `k (theta-theta0)^2`. I hadn't added the code to correctly handle these terms, so I had to add that too ;-)

Sire can now correctly read and write this file (using the code in the [fix_26 branch](https://github.com/OpenBioSim/sire/tree/fix_26). I still need to do some more work though, as at the moment this improper function type isn't supported by sire's internal molecular mechanics engine. I want to add this functionality before merging so that I can runs tests that verify the file is being parsed correctly by comparing energies.

Once this is done I will raise a pull request. We have missed the window for `2023.2.0`, but this should be able to go into `2023.2.1` which will be released likely mid to late April. Or, once merged, this will be in the development conda package from the `dev` channel.

**Comment by chryswoods** [comment]
I've merged the code into `devel`. It will take a few hours to build and deploy to anaconda cloud (as the latest `dev` channel package). This will let you load and save your molecule in gromacs topology format, but won't yet translate to Amber prmtop or calculate internal energies within sire. This should all work with BioSimSpace if you are loading files, then running simulations using the Gromacs engine.

We'll add the code to convert to other file formats and support internal energy calculations as part of a future feature branch which will be released in `2023.3.0` (due to be released at the end of June).

**Comment by ale94mleon** [comment]
Thanks! It sounds great.

**Comment by chryswoods** [comment]
I've renamed this to [feature_gromacs_energy](https://github.com/OpenBioSim/sire/tree/feature_gromacs_energy) and have create this issue (#39) to discuss and track that development (linked to the [2023.3.0 milestone](https://github.com/OpenBioSim/sire/milestone/1)). I'll close this issue as we will migrate any further conversation to there.

Thanks again - this is a really useful feature to add :-)
